src_features/signMessageEIP712/filtering.c

```diff
@@ -0,0 +1,185 @@
+                        hash_nbytes((uint8_t *) ".[]", 3, hash_ctx);
+                    }
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Verify the provided signature
+ *
+ * @param[in] dname_length length of provided substitution name
+ * @param[in] dname provided substitution name
+ * @param[in] sig_length provided signature length
+ * @param[in] sig pointer to the provided signature
+ * @param[in] type the type of filtering
+ * @return whether the signature verification worked or not
+ */
+static bool verify_filtering_signature(uint8_t dname_length,
+                                       const char *const dname,
+                                       uint8_t sig_length,
+                                       const uint8_t *const sig,
+                                       e_filtering_type type) {
+    uint8_t hash[INT256_LENGTH];
+    cx_ecfp_public_key_t verifying_key;
+    cx_sha256_t hash_ctx;
+    uint64_t chain_id;
+
+    cx_sha256_init(&hash_ctx);
+
+    // Magic number, makes it so a signature of one type can't be used as another
+    switch (type) {
+        case FILTERING_SHOW_FIELD:
+            hash_byte(FILTERING_MAGIC_STRUCT_FIELD, (cx_hash_t *) &hash_ctx);
+            break;
+        case FILTERING_PROVIDE_MESSAGE_INFO:
+            hash_byte(FILTERING_MAGIC_CONTRACT_NAME, (cx_hash_t *) &hash_ctx);
+            break;
+        default:
+            apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+            PRINTF("Invalid filtering type when verifying signature!\n");
+            return false;
+    }
+
+    // Chain ID
+    chain_id = __builtin_bswap64(eip712_context->chain_id);
+    hash_nbytes((uint8_t *) &chain_id, sizeof(chain_id), (cx_hash_t *) &hash_ctx);
+
+    // Contract address
+    hash_nbytes(eip712_context->contract_addr,
+                sizeof(eip712_context->contract_addr),
+                (cx_hash_t *) &hash_ctx);
+
+    // Schema hash
+    hash_nbytes(eip712_context->schema_hash,
+                sizeof(eip712_context->schema_hash),
+                (cx_hash_t *) &hash_ctx);
+
+    if (type == FILTERING_SHOW_FIELD) {
+        hash_filtering_path((cx_hash_t *) &hash_ctx);
+    } else  // FILTERING_PROVIDE_MESSAGE_INFO
+    {
+        hash_byte(ui_712_remaining_filters(), (cx_hash_t *) &hash_ctx);
+    }
+
+    // Display name
+    hash_nbytes((uint8_t *) dname, sizeof(char) * dname_length, (cx_hash_t *) &hash_ctx);
+
+    // Finalize hash
+    cx_hash((cx_hash_t *) &hash_ctx, CX_LAST, NULL, 0, hash, INT256_LENGTH);
+
+    cx_ecfp_init_public_key(CX_CURVE_256K1,
+                            LEDGER_SIGNATURE_PUBLIC_KEY,
+                            sizeof(LEDGER_SIGNATURE_PUBLIC_KEY),
+                            &verifying_key);
+    if (!cx_ecdsa_verify(&verifying_key, CX_LAST, CX_SHA256, hash, sizeof(hash), sig, sig_length)) {
+#ifndef HAVE_BYPASS_SIGNATURES
+        PRINTF("Invalid EIP-712 filtering signature\n");
+        apdu_response_code = APDU_RESPONSE_INVALID_DATA;
```

Comment:
That's right, the whole APDU exchange for an EIP712 message signature is complex and the python Ethereum client only produces valid APDUs.
It would however be possible to send the APDU out of order, but sending straight-up incorrect ones, not possible and I've not found a way so far to do it elegantly.